### PR TITLE
trf 4.09

### PR DESCRIPTION
**Pull request by tseemann**
### Have you:

- [x] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [x] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?
- [x] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [x] Built your formula locally prior to submission with `brew install <formula>`?

### New formula: have you

- [x] Written a sensible test? (The best test is to compile and run a sample program.)

### Updates to existing formula: have you

- [x] Removed the `revision` line, if any, when bumping the version number?
- [x] Not altered the `bottle` section?
- [x] Checked that the tests still pass?

### CODE REVIEW

**Comment by tseemann** [comment]
If this works, I count it in my top 10 formula successes! 🙏

File: trf.rb [hunk 1]
```diff
@@ -1,28 +1,49 @@
...
   # The license does not permit redistribution. http://tandem.bu.edu/trf/trf.license.html
   bottle :unneeded
 
+  depends_on "patchelf" => :build unless OS.mac?
+
   def install
-    bin.install "trf.download.pl" => "trf"
+    exe = "trf.download.pl"
+    chmod 0755, exe
+    bin.install exe => "trf"
```
Comment (sjackman) [inline trf.rb]: Please revert this chunk. `brew` should automatically set the executable bit after installing of all ELF files.

File: trf.rb [hunk 2]
```diff
@@ -1,28 +1,49 @@
...
+    (testpath/"test.fa").write <<-EOS.undent
+       >seq
+       aggaaacctgccatggcctcctggtgagctgtcctcatccactgctcgctgcctctccag
+       atactctgacccatggatcccctgggtgcagccaagccacaatggccatggcgccgctgt
+       actcccacccgccccaccctcctgatcctgctatggacatggcctttccacatccctgtg
+    EOS
+    assert_match version.to_s, shell_output("#{bin}/trf #{testpath}/test.fa 2 7 7 80 10 50 500 2>&1", 1)
+    out = "#{testpath}/test.fa.2.7.7.80.10.50.500.1.txt.html"
+    assert File.exist?(out)
+    assert_match "Length: 180", File.read(out)
```
Comment (sjackman) [inline trf.rb]: Thanks for improving the test!

File: trf.rb [hunk 3]
```diff
@@ -1,28 +1,49 @@
...
+    exe = "trf.download.pl"
+    chmod 0755, exe
+    bin.install exe => "trf"
+    # as this is a binary, ensure it is using brewed libm, libc etc.
+    unless OS.mac?
+      system "patchelf",
+        "--set-rpath", HOMEBREW_PREFIX,
+        "--set-interpreter", HOMEBREW_PREFIX/"lib/ld.so",
+        bin/"trf"
+    end
```
Comment (sjackman) [inline trf.rb]: Nice fix! That likely explains the weird `ldd` failure that we've been seeing, if the host wasn't able to run the binary executable, due to too old `glibc` or something similar.

File: trf.rb [hunk 4]
```diff
@@ -1,28 +1,49 @@
 class Trf < Formula
   desc "Tandem repeats finder"
   homepage "https://tandem.bu.edu/trf/trf.html"
-  # doi '10.1093/nar/27.2.573'
+  # doi "10.1093/nar/27.2.573"
   # tag "bioinformatics"
-  version "4.07b"
+  version "4.09"
```
Comment (sjackman) [inline trf.rb]: Please reorder these lines to `desc homepage url sha256 version doi tag`.

**Comment by tseemann** [comment]
@sjackman I made the changes you said, and now i'm getting the 
```
brew reinstall -s -v /home/tseemann/git/homebrew-science/trf.rb
rm /home/linuxbrew/.linuxbrew/bin/trf
==> Reinstalling trf
==> Downloading https://tandem.bu.edu/cgi-bin/trf/trf.download.pl?fileselect=30
Already downloaded: /home/linuxbrew/.cache/Homebrew/trf-4.09.pl
==> Verifying trf-4.09.pl checksum
==> patchelf --set-rpath /home/linuxbrew/.linuxbrew --set-interpreter /home/linuxbrew/.linuxbrew/lib/ld.so /home/linuxbrew/.linuxbrew/Cellar/trf/4.09/bin/trf
ln -s ../Cellar/trf/4.09/bin/trf trf
Error: Failure while executing: /home/linuxbrew/.linuxbrew/opt/glibc/bin/ldd /home/linuxbrew/.linuxbrew/Cellar/trf/4.09/bin/trf
/home/linuxbrew/.linuxbrew/opt/glibc/bin/ldd: line 117: /home/linuxbrew/.linuxbrew/Cellar/trf/4.09/bin/trf: Permission denied
```

File: trf.rb [hunk 5]
```diff
@@ -1,28 +1,49 @@
...
 
   if OS.mac?
-    url "https://tandem.bu.edu/cgi-bin/trf/trf.download.pl?fileselect=29"
-    sha256 "385ef13d0fcf53532fc0bdfe386c6789323a144b95c0e4b42c54b382bc5a7672"
+    url "https://tandem.bu.edu/cgi-bin/trf/trf.download.pl?fileselect=32"
```
Comment (sjackman) [inline trf.rb]: Try changing this to http://tandem.bu.edu/trf/downloads/trf409.macosx

File: trf.rb [hunk 6]
```diff
@@ -1,28 +1,49 @@
...
+    sha256 "383213647a8ddeb08380d2a80bb265d16c925ef16a353c999f811ea5ea6212f8"
   elsif OS.linux?
-    url "https://tandem.bu.edu/cgi-bin/trf/trf.download.pl?fileselect=26"
-    sha256 "a3a760c7b74c9603fbc08d95e8fa696c00f35a2f179b0bd63b2b13757ad3b471"
+    url "https://tandem.bu.edu/cgi-bin/trf/trf.download.pl?fileselect=30"
```
Comment (sjackman) [inline trf.rb]: Try changing this to http://tandem.bu.edu/trf/downloads/trf409.linux64

**Comment by sjackman** [comment]
Okay. I've finally figured this issue out. It's a bug in `brew`. `brew` checks the library linkage of the ELF files before it runs the cleaner that sets the executable bit on those ELF files. Additionally, the cleaner is run only on `bin`, `sbin`, and `lib`. `ldd` fails if it's run on an ELF file that does not have its executable bit set. I've opened PR https://github.com/Linuxbrew/brew/pull/476 to address this issue.

This issue was made difficult to troubleshoot due to `brew` suppressing the `stderr` output of `ldd`. I've opened PR https://github.com/Homebrew/brew/pull/3099 to address that separate issue. `ldd` was outputting the following useful error message to stderr, which was being suppressed by https://github.com/Homebrew/brew/pull/3099/files#diff-83e8c13ed518191365f4edaa7339293dL16
```
ldd: warning: you do not have execution permission for `/home/sjackman/.linuxbrew/Cellar/trf/4.09/bin/trf'
```

**Comment by sjackman** [comment]
I've merged PR https://github.com/Linuxbrew/brew/pull/476, so the issue of `trf` not being executable causing `ldd` to fail should be resolved.

**Comment by tseemann** [comment]
@sjackman me being stuck with RHEL reveals so many subtle things!

File: trf.rb (hunk 5 above)
Comment (tseemann) [reply trf.rb]: Nice find.

File: trf.rb (hunk 6 above)
Comment (tseemann) [reply trf.rb]: Yep, i changed to https too.

**Pull request merged**